python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py

```diff
@@ -289,9 +290,10 @@ def fit(self,
         init_params.update(self.worker_init_params)
 
         # Check uniqueness of the MainCallback
-        if not callbacks:
-            callbacks = []
+        callbacks = callbacks or []
         make_only_mainCallback(callbacks)
+        if self.use_tqdm:
+            callbacks.append(TqdmCallback())
```

Comment:
Do we need to check if Tqdm is already in the callback (explicitly written by users, not recommended, but they can)
Or does it have issue when there are more than 1 Tqdm callback?

Reply:
> Do we need to check if Tqdm is already in the callback (explicitly written by users, not recommended, but they can) Or does it have issue when there are more than 1 Tqdm callback?

There will be two or more progress bars if so? So this could be sticked as an issue.

Reply:
Or more generally, we need to better handle such cases:
- If users wrongly input two identical callbacks, we need to correctly handle this (probably just ignore one or raise an error)
- From our side, if users want to write their custom callback, they are supposed to only inherit MainCallBack right? Or they have the flexibility to overwrite other callbacks? In the latter case, we may need to consider whether we suggest them to write everything in one callback or split them according to piece-wise functions.

Reply:
> Or more generally, we need to better handle such cases:
> 
> * If users wrongly input two identical callbacks, we need to correctly handle this (probably just ignore one or raise an error)

Yes, this could be fixed.

> * From our side, if users want to write their custom callback, they are supposed to only inherit MainCallBack right? Or they have the flexibility to overwrite other callbacks? In the latter case, we may need to consider whether we suggest them to write everything in one callback or split them according to piece-wise functions.

Users are allowed to write hooks more than only MainCallBack, however it would be easier if they write down all the logic they expect in MainCallBack. In my imagation it is more like our job to supply small hooks to achieve piece-wise functions?

Reply:
Then to make everything under our control, in our document, we just say users can only override MainCallBack?

Reply:
> Then to make everything under our control, in our document, we just say users can only override MainCallBack?


For most of the cases if users aren't going to modify a small part of hook, then all in one maincallback would be nice, but we may just **strongly suggest** them to write codes in only one maincallback